Sources/SwiftGRPC/Core/Channel.swift

```diff
@@ -81,4 +81,145 @@ public class Channel {
     let underlyingCall = cgrpc_channel_create_call(underlyingChannel, method, host, timeout)!
     return Call(underlyingCall: underlyingCall, owned: true, completionQueue: completionQueue)
   }
+
+  public func connectivityState(tryToConnect: Bool = false) -> ConnectivityState {
+    return ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(underlyingChannel, tryToConnect ? 1 : 0))
+  }
+
+  public func subscribe(sourceState: ConnectivityState, tryToConnect: Bool = false, callback: @escaping (ConnectivityState) -> ()) {
+    var observer = observers.first(where: { $0.state == sourceState })
+
+    if observer == nil {
+      let newObserver = ConnectivityObserver(state: sourceState, underlyingChannel: underlyingChannel, tryToConnect: tryToConnect)
+      observers.append(newObserver)
+      observer = newObserver
+    }
+
+    observer?.callbacks.append(callback)
+    observer?.polling = true
+  }
+}
+
+private extension Channel {
+  class ConnectivityObserver: Equatable {
+    let state: ConnectivityState
+    let queue: CompletionQueue
+    let underlyingChannel: UnsafeMutableRawPointer
+    let underlyingCompletionQueue: UnsafeMutableRawPointer
+    private(set) var tryToConnect: Bool
+    var callbacks: [(ConnectivityState) -> ()] = []
+    private var lastState: ConnectivityState
+
+    var polling: Bool = false {
+      didSet {
+        if polling == true && oldValue == false {
+          run()
+        }
+      }
+    }
+
+    init(state: ConnectivityState, underlyingChannel: UnsafeMutableRawPointer, tryToConnect: Bool) {
+      self.state = state
+      self.underlyingChannel = underlyingChannel
+      self.tryToConnect = tryToConnect
+      self.underlyingCompletionQueue = cgrpc_completion_queue_create_for_next()
+      self.queue = CompletionQueue(underlyingCompletionQueue: self.underlyingCompletionQueue, name: "Connectivity State")
+      self.lastState = ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(self.underlyingChannel, 0))
+    }
+
+    deinit {
+      queue.shutdown()
+    }
+
+    private func run() {
+      DispatchQueue.global().async { [weak self] in
+        guard let `self` = self, let underlyingState = self.lastState.underlyingState else { return }
+
+        while self.polling {
+          guard !self.callbacks.isEmpty && !self.tryToConnect else {
+            self.polling = false
+            break
+          }
+
+          defer { self.tryToConnect = false }
+
+          let deadline: TimeInterval = 0.2
+          cgrpc_channel_watch_connectivity_state(self.underlyingChannel, self.underlyingCompletionQueue, underlyingState, deadline, nil)
```

Comment:
Shouldn't `underlyingState` be the last observed state? At the moment this is only updated once, at the very beginning of the spinloop, so this will fire for any states that are different from the state when the loop was started, instead of simply firing every time we detect any change.